espnet2/asr_transducer/decoder/stateless_decoder.py

```diff
@@ -0,0 +1,144 @@
+"""Stateless decoder definition for Transducer models."""
+
+from typing import List, Optional, Tuple, Union
+
+import torch
+from typeguard import check_argument_types
+
+from espnet2.asr_transducer.beam_search_transducer import ExtendedHypothesis, Hypothesis
+from espnet2.asr_transducer.decoder.abs_decoder import AbsDecoder
+
+
+class StatelessDecoder(AbsDecoder):
+    """Stateless Transducer decoder module.
+
+    Args:
+        vocab_size: Output size.
+        embed_size: Embedding size.
+        embed_dropout_rate: Dropout rate for embedding layer.
+        embed_pad: Embed/Blank symbol ID.
+
+    """
+
+    def __init__(
+        self,
+        vocab_size: int,
+        embed_size: int = 256,
+        embed_dropout_rate: float = 0.0,
+        embed_pad: int = 0,
+    ) -> None:
+        assert check_argument_types()
+
+        super().__init__()
+
+        self.embed = torch.nn.Embedding(vocab_size, embed_size, padding_idx=embed_pad)
+        self.embed_dropout_rate = torch.nn.Dropout(p=embed_dropout_rate)
+
+        self.output_size = embed_size
+        self.vocab_size = vocab_size
+
+        self.blank_id = embed_pad
+
+        self.device = next(self.parameters()).device
+        self.score_cache = {}
+
+    def forward(
+        self,
+        labels: torch.Tensor,
+        states: Optional[Tuple[torch.Tensor, Optional[torch.Tensor]]] = None,
+    ) -> torch.Tensor:
+        """Encode source label sequences.
+
+        Args:
+            labels: Label ID sequences. (B, L)
+            states: Decoder hidden states. None
+
+        Returns:
+            dec_embed: Decoder output sequences. (B, U, D_emb)
+
+        """
+        dec_embed = self.embed_dropout_rate(self.embed(labels))
+
+        return dec_embed
+
+    def score(
+        self,
+        label: torch.Tensor,
+        label_sequence: List[int],
+        state: None,
+    ) -> Tuple[torch.Tensor, None]:
+        """One-step forward hypothesis.
+
+        Args:
+            label: Previous label. (1, 1)
+            label_sequence: Current label sequence.
+            state: Previous decoder hidden states. None
+
+        Returns:
+            dec_out: Decoder output sequence. (1, D_emb)
+            state: Decoder hidden states. None
+
+        """
+        str_labels = "_".join(map(str, label_sequence))
+
+        if str_labels in self.score_cache:
+            dec_embed = self.score_cache[str_labels]
+        else:
+            dec_embed = self.embed(label)
+
+            self.score_cache[str_labels] = dec_embed
+
+        return dec_embed[0], None
+
+    def batch_score(
+        self,
+        hyps: Union[List[Hypothesis], List[ExtendedHypothesis]],
```

Comment:
`ExtendedHypothesis` is a subclass of `Hypothesis`, so it is safe to use
```python3
hyps: List[Hypothesis]
```
here. No need to use 
`Union[List[Hypothesis], List[ExtendedHypothesis]]`